### DISCUSSION TITLE
Setup 2.8 inch ST7789 / ILI9341 screen

### DISCUSSION THREAD

**Discussion by salarizadi** [Moddable-OpenSource/moddable#1175] [Q&A] [answered]
Hi.

Module : ESP32-WROOM-32

Display screen : https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/displays/wiring-guide-generic-2.4-spi-esp32.md

Driver : ST7789 - v1.1

Manifest : 
```json
{
			"include": [
				"$(MODULES)/drivers/st7789/manifest.json",
				"$(MODDABLE)/modules/drivers/xpt2046/manifest.json"
			],
			"defines": {
				"spi": {
					"mosi_pin": 22,
					"miso_pin": 21,
					"sck_pin": 23
				},
				"ili9341": {
					"cs_pin": 18,
					"dc_pin": 19,
					"rst_pin": 5,
					"touch_pin": 4,
					"spi_port": "HSPI_HOST",
					"flipX": true
				},
				"xpt2046": {
					"cs_pin": 4,
					"touch_pin": 15,
					"spi_port": "HSPI_HOST",
					"raw_left": 141,
					"raw_right": 1955,
					"raw_top": 1970,
					"raw_bottom": 158
				}
			}
		}
```

I used the drag and drop piu example, but unfortunately, when dragging the objects, it doesn't recognize the touch properly, and for example, if I touch the top of the screen, it thinks I'm touching the bottom of the screen. How should I solve this problem?

**Comment by salarizadi** [comment]
When drag objects :

![20230716_205219.jpg](https://github.com/Moddable-OpenSource/moddable/assets/67143370/3a6544b6-a56c-4869-953d-f6e6c22012ff)

**Comment by salarizadi** [comment]
Why?!!

https://github.com/Moddable-OpenSource/moddable/assets/67143370/31d4980c-e34b-4c7d-a733-d604b6467daa

**Reply by salarizadi** [reply, marked as answer]
If you also had this problem and could not receive the touch correctly and the opposite points were touched, it is enough to solve this problem in xpt2046 definitions:

Instead :

```json
 "xpt2046": {
      "raw_left": 141,
      "raw_right": 1955,
      "raw_top": 1970,
      "raw_bottom": 158
 }
```

write this:

```json
 "xpt2046": {
      "raw_left": 1955,
      "raw_right": 141,
      "raw_top": 158,
      "raw_bottom": 1970
 }
```

**Reply by phoddie** [reply]
For the XPS2046 driver, you can run the [calibration example](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/drivers/xpt2046calibrate/main.js) to adjust the touch points. They can be a little different on each display.

**Reply by salarizadi** [reply]
> For the XPS2046 driver, you can run the [calibration example](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/drivers/xpt2046calibrate/main.js) to adjust the touch points. They can be a little different on each display.

That's interesting, thanks, I'll definitely have to test this out

**Reply by salarizadi** [reply]
@phoddie I tried the calibration and put the exact numbers, unfortunately it didn't work and it has exactly the same problem as the horizontal keyboard, but the vertical one works without any problem. Now I want to test the horizontal calibration, but since I have never worked with Poco before, I don't know where and how to rotate the screen. The only thing I saw in the document but I didn't understand where to use it is kPocoRotation?


```json
{
  "include": [
    "$(MODDABLE)/examples/manifest_base.json",
    "$(MODDABLE)/examples/manifest_piu.json",
    "$(MODULES)/input/keyboard/manifest.json"
  ],
  "modules": {
    "*": [
      "./main"
    ]
  },
  "resources": {
    "*-alpha": [
      "$(MODDABLE)/examples/assets/fonts/OpenSans-Regular-20",
      "$(MODDABLE)/examples/assets/fonts/OpenSans-Semibold-18"
    ]
  },
  "creation": {
    "stack": 350
  },
  "config": {
    "format": "RGB565LE",
    "touchCount": 1
  },
  "platforms": {
    "esp32": {
      "include": [
        "$(MODULES)/drivers/st7789/manifest.json",
        "$(MODDABLE)/modules/drivers/xpt2046/manifest.json"
      ],
      "config": {
        "screen": "st7789",
        "touch": "xpt2046",
        "rotation": 270
      },
      "defines": {
        "spi": {
          "mosi_pin": 22,
          "miso_pin": 21,
          "sck_pin": 23
        },
        "ili9341": {
          "cs_pin": 18,
          "dc_pin": 19,
          "rst_pin": 5,
          "hz": 25000000,
          "spi_port": "HSPI_HOST",
          "width": 240,
          "height": 320
        },
        "xpt2046": {
          "cs_pin": 4,
          "touch_pin": 15,
          "hz": 4000000,
          "spi_port": "HSPI_HOST",
          "raw_left": 1908,
          "raw_right": 129,
          "raw_top": 100,
          "raw_bottom": 1861
        }
      }
    }
  }
}

```

**Comment by salarizadi** [comment]
Why is the color different??!

PIU transition :

![20230716_214946.jpg](https://github.com/Moddable-OpenSource/moddable/assets/67143370/fa5afc38-4fa8-4feb-8425-005134cbcae2)

**Reply by salarizadi** [reply]
This is not a bug, but it is accidentally listed in the ST7789 register and the 0x21 command is to turn on invert. If you encounter this problem with the command

```js
screen.command(0x20) // invert off
```

Solve the problem.

**Reply by phoddie** [reply]
Different displays require different register configurations. That is why the driver allows you to specify the [registers](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/drivers/ili9341/ili9341.md#defines) in your manifest.
